pkg/codec/byte_string_modifier.go

```
@@ -0,0 +1,307 @@
+		}
+		return nil, fmt.Errorf("%w: cannot convert bytes for field %s", types.ErrInvalidType, field)
+
+	default:
+		return nil, fmt.Errorf("%w: cannot convert bytes for field %s", types.ErrInvalidType, field)
+	}
+}
+
+// stringToAddressHookForOnChain converts a string representation of an address back into a byte array for on-chain use.
+func stringToAddressHookForOnChain(modifier AddressModifier) func(from reflect.Type, to reflect.Type, data any) (any, error) {
+	return func(from reflect.Type, to reflect.Type, data any) (any, error) {
+		byteArrTyp := reflect.ArrayOf(modifier.Length(), reflect.TypeOf(byte(0)))
+		strTyp := reflect.TypeOf("")
+
+		// Convert from string to byte array (e.g., string -> [20]byte)
+		if from == strTyp && (to == byteArrTyp || to.ConvertibleTo(byteArrTyp)) {
+			addr := data.(string)
+
+			// Decode the address based on encoding type
+			bts, err := modifier.DecodeAddress(addr)
+			if err != nil {
+				return nil, err
+			}
+
+			// Ensure the byte length matches the expected length
+			if len(bts) != modifier.Length() {
+				return nil, fmt.Errorf("length mismatch: expected %d bytes, got %d", modifier.Length(), len(bts))
+			}
+
+			// Create a new array of the desired type and fill it with the decoded bytes
+			val := reflect.New(byteArrTyp).Elem()
+			reflect.Copy(val, reflect.ValueOf(bts))
+			return val.Interface(), nil
+		}
+		return data, nil
+	}
+}
+
+// addressToStringHookForOffChain converts byte arrays to their string representation for off-chain use.
+func addressToStringHookForOffChain(modifier AddressModifier) func(from reflect.Type, to reflect.Type, data any) (any, error) {
+	return func(from reflect.Type, to reflect.Type, data any) (any, error) {
+		byteArrTyp := reflect.ArrayOf(modifier.Length(), reflect.TypeOf(byte(0)))
+		strTyp := reflect.TypeOf("")
+
+		// if 'from' is a pointer to the byte array (e.g., *[20]byte), dereference it.
+		if from.Kind() == reflect.Ptr && from.Elem() == byteArrTyp {
```

I don't 100% remember how this works, but I think we can return an error if fields aren't what we are expecting. This would mean that someone defined an unnecessary modifier right?

---

It was in case we received a pointer to the array. I think we can just grab the underlying value and proceed

---

I'm saying that if `from` is not of any type that we are expecting should we return an error?

---

Realized this is not needed so I'll just remove it